lib/logstash/inputs/udp.rb

```diff
@@ -147,8 +150,12 @@ def inputworker(number, codec)
         break if payload == :END
 
         ip_address = client[3]
-
-        codec.decode(payload) { |event| push_decoded_event(ip_address, event) }
+        if @metadata
+          metadata_to_codec = {}
+          metadata_to_codec["port"] = client[1]
+          metadata_to_codec["host"] = client[3]
+        end
+        codec.decode(payload, metadata_to_codec) { |event| push_decoded_event(ip_address, event) }
```

Comment:
what if `metadata_to_codec` is not defined here?

Reply:
I understand this defaults to nil from parent.

Reply:
The public API for `LogStash::Codecs::Base#decode` accepts a single argument, so if we send two arguments to any plugin that doesn't explicitly override the base implementation this will fail with an `ArgumentError`.


I'm not sure if the following is clear enough, but it will populate `optional_metadata` as an array containing one or zero items based on the value of `@metadata`, and then "splat" that array out as arguments for decode (causing two arguments to be sent when `@metadata` is true and only one to be sent when `@metadata` is false)

~~~ ruby
     optional_metadata = @metadata ? [{"port" => client[1], "host" => client[3]}] : []
     codec.decode(payload, *optional_metadata) { |event| push_decoded_event(ip_address, event) }
~~~

Reply:
another option would be to define a new method `codec_decode` that handles this

~~~ ruby
    ##
    # @param codec [LogStash::Codecs::Base]
    # @param data [String]
    # @param client [Array(Void, Integer,Void,String)]
    # @yieldparam [LogStash::Event]
    # @yieldreturn [Void]
    # @return void
    def codec_decode(codec, payload, client, &block)
      if @metadata
        connection_metadata = {"host"=>client[3], "port" => client[1]}
        codec.decode(payload, connection_metadata, &block)
      else
        codec.decode(payload, &block)
      end
    end
~~~

Which we could then use here:

```suggestion
        codec_decode(codec, payload, client) { |event| push_decoded_event(ip_address, event) }
```

Reply:
Thanks for your comments and effort, much appreciated. I'm happy to provide this as an idea for further development.

Reply:
@yaauie Thanks for your feedback, marking this as resolved.